src/google/protobuf/compiler/cpp/cpp_helpers.h

```diff
@@ -746,25 +746,19 @@ class PROTOC_EXPORT NamespaceOpener {
   void ChangeTo(const std::string& name) {
     std::vector<std::string> new_stack_ =
         Split(name, "::", true);
-    int len = std::min(name_stack_.size(), new_stack_.size());
-    int common_idx = 0;
-    while (common_idx < len) {
-      if (name_stack_[common_idx] != new_stack_[common_idx]) break;
-      common_idx++;
-    }
-    for (int i = name_stack_.size() - 1; i >= common_idx; i--) {
-      if (name_stack_[i] == "PROTOBUF_NAMESPACE_ID") {
+    for (auto oldNameIt = name_stack_.crbegin(); oldNameIt != name_stack_.crend(); ++oldNameIt) {
```

Comment:
The index-based algorithm version used before turned out to be unstable to replacing int with size_t.
Hence I have changed it to range-based loops (C++ does not have reversed range-based for yet, hence iterators were used).

The new algorithm fully closes nested namespaces and fully opens them.

As of C++17 we have nested namespaces, so none of this will be required soon.

Reply:
Won't this new version result in more namespace openings and closings than necessary? I thought a goal of the old version was to figure out the longest common namespace prefix so that we don't have to close and reopen the same namespaces unnecessarily. E.g. if we're changing from `foo::bar` to `foo::baz` then we don't need to close and reopen `foo`.

Reply:
Oh, I see you already mentioned this in your comment. I think we probably need to keep the old behavior unfortunately. It might be a while before we can start using the C++17 syntax because we don't have any plans to require C++14 or C++17 yet.

Reply:
Indeed, the number of namespace operations was increaced in this PR, but I do not think it will lead to any observable change (please, tell if you know of any).

I have attempted to fix -Wsign-compare without simplifying the code around, but proper fix would require more obfuscation: it looks like the current version attempts to operate negative indices in 
`for (int i = name_stack_.size() - 1; i >= common_idx; i--) { ` 

If I change i to being unsigned, an overflow will be triggered.

Reply:
There wouldn't be an observable semantic change, but I think there could be a build performance regression just from the increase in the generated source code size. I believe within Google we have some proto files that could have pb.cc files thousands of lines longer if we fully closed and reopened the namespaces.

I'm not sure what the best solution for the compiler warnings would be, but maybe one easy fix would be to just keep everything signed and immediately `static_cast` to `int` all the `size()` calls.

Reply:
Ok, I will try to find a better solution that would not cause additional namespace operations.

Reply:
I have restored the old behavior, but I keep using iterators for reversed iteration during namespace closure.